enhancements/sig-architecture/58-addon-placement-annotation/README.md

```diff
@@ -0,0 +1,65 @@
+2. Tolerations annotation, the key of this annotation is: `addon.open-cluster-management.io/tolerations`, the value of this annotation represents a toleration array with json format
+
+There is an example
+
+```yaml
+apiVersion: addon.open-cluster-management.io/v1alpha1
+kind: ManagedClusterAddOn
+metadata:
+  name: work-manager
+  namespace: cluster1
+  annotations:
+    addon.open-cluster-management.io/nodeSelector: "{\"kubernetes.io/os\":\"linux\"}"
+    addon.open-cluster-management.io/tolerations: "[{\"key\":\"node.kubernetes.io/unreachable\",\"operator\":\"Exists\",\"effect\":\"NoSchedule\"}]"
+spec:
+  installNamespace: open-cluster-management-agent-addon
+```
+
+### Design Details
+
+Currently, there is a function [`GetValuesFromAddonAnnotation`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/helper.go#L21) in the addon-framework, it returns a [`Values`]([https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L22) object (the object type is `map[string]interface{}`) that is built from the addon values annotation, this object is used to render the addon deployment.
+In this proposal, if one addon has the node selector and tolerations annotations, this function will merge them to the `Values` object with the key `NodeSelector` and `Tolerations`.
+
+For an addon implementation, developers use [`AgentAddonFactory`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L28) to build their addon, they use [`WithGetValuesFuncs`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L61) to specify the get values functions (the function type is (`GetValuesFunc`)[https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L24]) to render their addon deployment, for get values function, the developers can choose to use addon-frameworks `GetValuesFromAddonAnnotation` function or implement it by themselves to get these two annotations.
+
+### Test Plan
+- test add placement annotation to an addon which has no values annotation 
+- test add placement annotation to an addon which has values annotation
+- test delete placement annotation from an addon which has no values annotation
+- test delete placement annotation from an addon which has values annotation
+
+### Graduation Criteria
+N/A
+
+### Upgrade Strategy
+
+If there is an existing addon CR that already used the addon values annotation to specify the node selector and tolerations and the addon implementation used the `GetValuesFromAddonAnnotation` function to get them. The addon user can choose to keep them in addon values annotation or use new addon placement annotation to define them and remove them from the addon values annotation.
+
+### Version Skew Strategy
+N/A
+
+## Alternatives
```

Comment:
yeah, we have considered this, if we add these to addon spec, it may not meet all needs, e.g. if one addon need to deploy more than one agent, and these agents have different node selector/tolerations, this will be hard to use

to keep the addon flexibility, we proposed this, providing an optional way for the addon developers, they can choose use this way or use their own way to handle the node selector/tolerations for their addons, e.g. the developers may have own addon configuration cr, and use the configuration to render their addon agents

Reply:
I think we cannot ensure the agent will use this even if adding this to spec. installNamespace is different, since it is needed for the registration controller to put secret for the addon to communicate to hub. @skeeey let's add spec as an atlernative in the proposal, and clarify the reason.

Reply:
added

Reply:
I had the same thought. I think the additional flexibility is a negative in this case. We want a consistent interface for how to configure addons from a user's perspective. If this is free-form JSON strings that are different per addon, I think this becomes hard to use and confusing. It's especially confusing when a user puts in an invalid JSON string since there is no immediate feedback from API validation. It also makes API compatibility hard since it becomes up to each OCM addon how to interpret that JSON strin.

I also dislike that an annotation has a JSON string as a value since these are not user-friendly to read or write and suggest that a new item in the spec is more appropriate.

My suggestion is to make the spec flexible enough for all current OCM addons and make it optional. Any complex configurations beyond this can be addon specific in their own CRDs.

Reply:
but nodeSelect/toleration is not a common property for every addon. Having nodeSelect/toleration in the spec gives an impression that the the addon will honor this field, but it is not true. If we want a more strict validation, I think we should have an Object ref field (@jnpacker mentioned this also) instead of putting nodeselector in the addon spec.

Reply:
What about an objectRef or list of objectRef's.  It does mean that each addon may not configure the same set of values or even resources, but if as @mprahl  suggests, you used a CRD for your addon, you could have basic validation, defaults, etc...

Also a ConfigMap or secret could be leveraged to try things out, and at least the format is well understood.

Reply:
I will add the objectRef to the enhancement as an optional proposal

Reply:
@qiujian16 and @jnpacker this sounds interesting. I hadn't heard about `objectRef` before and couldn't find anything concrete online. Could you please explain what this concept is?

Reply:
@mprahl probably this https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5622-L5639